src/v0.rs

```diff
@@ -953,27 +956,30 @@ impl<'a, 'b, 's> Printer<'a, 'b, 's> {
         }
 
         let ty_tag = parse!(self, next);
+
+        if ty_tag == b'p' {
+            // We don't encode the type if the value is a placeholder.
+            self.out.write_str("_")?;
+            return Ok(());
+        }
+
         if !supported_const_generic_type(ty_tag) {
             invalid!(self);
         }
 
-        if self.eat(b'p') {
-            self.out.write_str("_")?;
-        } else {
-            match ty_tag {
-                // Unsigned integer types.
-                b'h' | b't' | b'm' | b'y' | b'o' | b'j' => self.print_const_uint()?,
-                // Signed integer types.
-                b'a' | b's' | b'l' | b'x' | b'n' | b'i' => self.print_const_int()?,
-                // Bool.
-                b'b' => self.print_const_bool()?,
-                // Char.
-                b'c' => self.print_const_char()?,
-
-                // This branch ought to be unreachable.
-                _ => invalid!(self),
-            };
-        }
+        match ty_tag {
+            // Unsigned integer types.
+            b'h' | b't' | b'm' | b'y' | b'o' | b'j' => self.print_const_uint()?,
+            // Signed integer types.
+            b'a' | b's' | b'l' | b'x' | b'n' | b'i' => self.print_const_int()?,
+            // Bool.
+            b'b' => self.print_const_bool()?,
+            // Char.
+            b'c' => self.print_const_char()?,
+
+            // This branch ought to be unreachable.
+            _ => invalid!(self),
+        };
```

Comment:
This can all be simplified by removing `supported_const_generic_type`.

I had some ideas for simplifying all of this code by removing the `skip_` vs `print_` split, given how much cleaner that made the C port, so I wouldn't worry about duplication.